src/Compilers/Core/Portable/Syntax/SyntaxNode.cs

```diff
@@ -426,24 +426,78 @@ public bool ContainsSkippedText
         }
 
         /// <summary>
-        /// Determines whether this node has any descendant preprocessor directives.
+        /// Determines whether this node or any of its descendant nodes, tokens or trivia have any diagnostics on them. 
         /// </summary>
-        public bool ContainsDirectives
+        public bool ContainsDiagnostics
         {
             get
             {
-                return this.Green.ContainsDirectives;
+                return this.Green.ContainsDiagnostics;
             }
         }
 
         /// <summary>
-        /// Determines whether this node or any of its descendant nodes, tokens or trivia have any diagnostics on them. 
+        /// Determines whether this node has any descendant preprocessor directives.
         /// </summary>
-        public bool ContainsDiagnostics
+        public bool ContainsDirectives => this.Green.ContainsDirectives;
+
+        /// <summary>
+        /// Returns true if this node contains any directives (e.g. <c>#if</c>, <c>#nullable</c>, etc.) within it with a matching kind.
+        /// </summary>
+        public bool ContainsDirective(int rawKind)
         {
-            get
+            var stack = PooledObjects.ArrayBuilder<GreenNode?>.GetInstance();
+            stack.Push(this.Green);
+
+            try
             {
-                return this.Green.ContainsDiagnostics;
+                while (stack.Count > 0)
+                {
+                    var current = stack.Pop();
+                    if (current is null)
+                        continue;
+
+                    // Don't bother looking further down this portion of the tree if it clearly doesn't contain directives.
+                    if (!current.ContainsDirectives)
+                        continue;
+
+                    if (current.IsToken)
+                    {
+                        // no need to look within if this token doesn't even have leading trivia.
```

Comment:
Is there a hard rule that directives are always leading trivia? Maybe because they cannot be preceded by non-trivia on a given line?

Reply:
yes.  they must always be leading since they must always be the first thing on a line.  trailing trivia always comes after a token on the same line, and just can never be directives.

Reply:
Might be worth asserting?